### ISSUE TITLE
Future of Device Class for Alexa UAs

### ISSUE THREAD

**Issue by benbeckingsale** [nielsbasjes/yauaa#211]
Hi Niels, I have a conceptual question around YAUAA’s classification of Amazon Alexa useragents, e.g.

    AlexaMediaPlayer/2.1.2860.0 (Linux;Android 5.1.1) ExoPlayerLib/1.5.9
    AlexaService/2.2.305814.0 (Linux;Android 9) ExoPlayerLib/2.8.4

Do you have any plans to change YAUAA’s classification of Device Classes for such UAs? It seems all are currently parsed as `DeviceClass: 'Mobile'`, but as Alexa comes to more and more devices, I am wondering how YAUAA might distinguish between ‘Voice’, ‘Mobile', and other device classes.
Any insight you could share would be gratefully appreciated.

**Comment by nielsbasjes** [comment]
These are now classified as "Mobile" because the only elements indicating the type of device are "Linux" and "Android". Because "Android" is (so far) mostly used on mobile devices and I cannot determine Phone, Tablet, Watch, ... it stays the (rather vague) "Mobile".

If there is something in the useragent that makes it more specific (like "Voice") for these systems I'm fine with putting a small table in that allows making this distinction better.

Do you have such an overview?
Preferably an official "spec" by the creators of these systems (Amazon?)

**Comment by benbeckingsale** [comment]
Hi Niels, many thanks for your response.
 
Firstly I should say that the AlexaService useragent is just an example; I am really only interested in AlexaMediaPlayer, which is very prevalent in my traffic.
 
There is this from [Amazon Developer's page on Alexa](https://developer.amazon.com/en-US/alexa):
 

> Alexa is Amazon’s cloud-based voice service available on hundreds of millions of devices from Amazon and third-party device manufacturers. With Alexa, you can build natural voice experiences that offer customers a more intuitive way to interact with the technology they use every day.

If Alexa is a voice-driven service, it seems accurate to classify devices running Alexa as 'voice-driven' devices. It would fit with the definition of `Device Class: Voice` in the YAUAA documentation: "a voice driven device (i.e. ask a question and the page is read aloud)"

In other words, I think the proof in the useragent is the presence of AlexaMediaPlayer; I would be confident that the device is voice-driven _because_ it made a request using Alexa. I believe voice is the only mode of interaction.
 
(As a side question, if a useragent was equally 'Voice' and 'Mobile' (e.g. from a mobile smart speaker), which class would you give precedence?)
 
Hopefully this makes sense. I would enjoy hearing your thoughts on whether this suggestion would improve YAUAA.

**Comment by nielsbasjes** [comment]
How can I distinguish between an "Alexa voice device" and (if it exists) an Alexa app on a phone?

**Comment by benbeckingsale** [comment]
[Alexa apps](https://www.amazon.co.uk/gp/help/customer/display.html?nodeId=GDHB98S32QPQ5H4Y) exist for iOS, Android and FireOS.
It looks like they are mostly intended for setting up Alexa devices, but also allow you to stream content and hand it off to Alexa devices, probably resulting in `AlexaMediaPlayer...`.

[The Amazon Music app](https://www.amazon.co.uk/b/ref=s9_acss_bw_cg_DMMAOMNV_1a1_w?node=14131688031&ref=dmm_lp_voice_nav_uk_new&pf_rd_m=A3P5ROKL5A1OLE&pf_rd_s=merchandised-search-2&pf_rd_r=NPNKNM60BJQYS73VV002&pf_rd_t=101&pf_rd_p=63df8d2b-9bdb-4e29-b2f3-ebeeaed7b03c&pf_rd_i=14142767031) lets you control playback on a mobile device using Alexa. I don't know what the useragent would be in this case, but I am not sure you could tell between it and one from an Alexa device consistently.

If the device class can't be distinguished, do you think `AgentClass: Voice` would be a more accurate assertion as Alexa is defined as a "voice service"?

**Comment by nielsbasjes** [comment]
I added the Alexa agents to https://github.com/nielsbasjes/yauaa/blob/AudioDevices/analyzer/src/main/resources/UserAgents/Audio.yaml
Appreciate it if you can verify if this is correct.

**Comment by benbeckingsale** [comment]
Thanks very much for adding Alexa to your test cases.

How do you differentiate `AgentClass` and `DeviceClass`? Perhaps one is software and the other hardware? After more thought I would agree with what you initially said: that the useragent indicates no device class other than `Mobile`. I'm sure "voice" is an attribute of the Alexa software instead of the hardware.

So to me it seems like the most accurate outcome for `AlexaMediaPlayer...` useragents is to keep `DeviceClass: Mobile` and change to `AgentClass: Voice`.

This might also be a neat distinction in a IoT future where many device types support Alexa.

Would you be willing to implement this?

**Comment by nielsbasjes** [comment]
I think this is what you meant: https://github.com/nielsbasjes/yauaa/blob/master/analyzer/src/main/resources/UserAgents/Audio.yaml#L279 can you please check this?
Thanks.

**Comment by benbeckingsale** [comment]
Hi Niels, thanks for your help with this request. The test case looks good to me.

Going forward I am going to continue testing useragent behaviour with different Alexa/Sonos configurations.
I have a suspicion that perhaps only Amazon devices produce the `AlexaMediaPlayer` useragent; if I find anything interesting I will raise a separate issue.

Thanks.

**Comment by nielsbasjes** [comment]
Cool. I'm looking forward to your findings. 
In the meantime I released version 5.18 that includes these rules.
